lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -2206,25 +2194,38 @@ static bool diagnoseUnsatisfiedRequirements(ADContext &context,
 /// tuple. Otherwise, add the value directly to `result`.
 static void collectAllExtractedElements(SILValue val,
                                         SmallVectorImpl<SILValue> &result) {
-  if (auto tupleType = val->getType().getAs<TupleType>()) {
-    result.resize(tupleType->getNumElements(), SILValue());
-    for (auto *use : val->getUses())
-      if (auto *tupleExtract = dyn_cast<TupleExtractInst>(use->getUser()))
-        result[tupleExtract->getFieldNo()] = tupleExtract;
-  }
-  else
+  auto tupleType = val->getType().getAs<TupleType>();
+  if (!tupleType) {
     result.push_back(val);
+    return;
+  }
+  result.reserve(tupleType->getNumElements());
+  bool visitedDTI = false;
+  for (auto *use : val->getUses()) {
+    if (auto *dti = dyn_cast<DestructureTupleInst>(use->getUser())) {
+      assert(!visitedDTI && "More than one 'destructure_tuple'!?");
```

Comment:
Did you mean to set `visitedDTI` to true within this loop?

Reply:
Yes, thanks!